### ISSUE TITLE
Failed to connect to server - HandshakeException

### ISSUE THREAD

**Issue by endervad** [UnicornsOnLSD/finamp#106]
I've set up a self-signed certificate on my Jellyfin server and installed it also as a trusted root CA on my devices. It works fine without security warnings on Chrome (both Windows and Android) and on Android apps, like official Jellyfin and Gelli, except Finamp. When I try to connect to my server in Finamp 0.5.1, I get
`HandshakeException: Handshake error in client (OS Error: CERTIFICATE_VERIFY_FAILED: unable to get local issuer certificate(handshake.cc:359))`

Certificate extensions:
`X509v3 extensions:
    X509v3 Key Usage: critical
        Digital Signature, Key Encipherment
    X509v3 Extended Key Usage:
        TLS Web Client Authentication, TLS Web Server Authentication
    X509v3 Subject Alternative Name:
        DNS:<redacted>
    X509v3 Basic Constraints: critical
        CA:TRUE
    X509v3 Subject Key Identifier:
        <redacted>`

**Comment by UnicornsOnLSD** [comment]
Try this APK, based off the allow-bad-certs branch: 
[app-arm64-v8a-release.zip](https://github.com/UnicornsOnLSD/finamp/files/7001443/app-arm64-v8a-release.zip)
[app-armeabi-v7a-release.zip](https://github.com/UnicornsOnLSD/finamp/files/7001445/app-armeabi-v7a-release.zip)
[app-x86_64-release.zip](https://github.com/UnicornsOnLSD/finamp/files/7001446/app-x86_64-release.zip)

This just allows all bad certificates, so I won't put this in the main branch. Just wondering, why don't you just use Let's Encrypt to get a proper certificate?

**Comment by endervad** [comment]
This one works, thanks :)
I have a probably rather unique case: I've got a DNS name to my public IP from ISP. It's not part of contract, and I'm not even sure that it was intended, but I don't pay for it so might as well use it as is.

**Comment by endervad** [comment]
Upon further usage of allow-bad-certs version it seems like I was careless to assume, that if I can see the library, everything works. Neither downloads, nor playing songs work.
In case of downloads there's only a single `PlatformException((0) Source error, null, null, null)` in Main Thread > AudioServiceHelper with null stack trace. And in playing there's the exact same handshake exception for every attempt to play any song in Audio Service > MusicPlayerBackgroundTask with null stack trace.

**Comment by UnicornsOnLSD** [comment]
Oh yeah, I forgot about native connections (the player and downloader are both native code). I'll look and see if there's any way of allowing bad certs.

About your setup though, are you connecting through an IP or a domain name? If it's a domain name, you should be able to get a certificate pretty easily.

**Comment by endervad** [comment]
It's a domain name, but not mine, it's ISP's. This same domain name is also used by ISP on their website, e.g. their website is example.com and DNS name to my assigned public IP is hostXXXX.example.com. Hence there's no way they would give me the private key to their certificate.

**Comment by UnicornsOnLSD** [comment]
You may be able to get a certificate for just your subdomain: https://stackoverflow.com/questions/39322112/multiple-subdomains-with-lets-encrypt

**Comment by endervad** [comment]
That would require action from ISP side, since they're the owner of the whole domain with all subdomains domains.
To get a certificate for a domain and/or subdomain you need to pass check(s) and prove that you're this domain owner. I'm not a domain owner in this case, I was just lucky enough to have a domain name for my IP. Ideally ISP could get a certificate specifically for my subdomain, yes, but there's no way they're going to do that, they're not that cooperative.

**Comment by UnicornsOnLSD** [comment]
Let's Encrypt's "most common" challenge type just requests a web path on your server with the certificate, no DNS required. I doubt it would work if your ISP has some kind of wildcard certificate for all domains though.

https://letsencrypt.org/docs/challenge-types/

**Comment by endervad** [comment]
Okay, somehow I was sure that DNS-01 is a "must" initially, and only then either DNS-01, or HTTP-01 is used for automatic renewing. My bad :D 
I was able to get a proper certificate, it's all good now on release version of Finamp. Sorry for taking up your time and thanks for pointing the challenge type stuff out :)
I can still see a scenario though, where allowing self-signed cert could be necessary. For example, if a person has public IP, doesn't have a domain name, and still wants to secure his traffic while not paying for a domain. It's possible to issue a certificate to an FQDN and then specify a desired IP in SAN extension. https://1.1.1.1 did it this way.

**Comment by Trinnik** [comment]
I have run into this problem with my setup at home and here's my scenario.

None of my services are publicly available, instead, I have an SDN setup using ZeroTier with an nginx proxy server that serves certificates locally. I have my root certificates installed on my user's devices so that they are able to have remote access including https with no open ports. This allows me the ability for my users to access everything as though it were on the public internet without having it actually available to the internet for security purposes.

**Comment by UnicornsOnLSD** [comment]
Im assuming ZeroTier works by you connecting to an IP? If you control a domain, you could use the DNS-01 challenge. If not, [it looks like ZeroTier's network is encrypted](https://en.m.wikipedia.org/wiki/ZeroTier#Security), however data may not be encrypted when travelling over the LAN.

I'll have a look at allowing custom certificates since I'm sure there are people out there who can't use a regular certificate, but last time I looked the only stuff I could find was just saying to disable certificate error handling, which isn't exactly secure.

**Comment by Trinnik** [comment]
ZeroTier uses peer to peer with root servers to negotiate the connection so unfortunately, everything on the network operates with virtual lan addresses' and node IDs, not publicly routeable.

I have no idea how you get access to the certificates trusted by the operating system but both Android and IOS allow you to trust foreign certificates. That's how I get my browsers to still tell me the webpage is secure. As long as my root certificate is installed at the OS level, it shows as a trusted cert.

I would assume that there's a way to access the trusted certificates from the OS. Love the app otherwise!

**Comment by sanyarnd** [comment]
Same thing here.

* I have a free DNS name from Mikrotik
* my ISP blocks 80 port

=> can't use HTTP/DNS challenge and have to use self-signed certificated :(

Would be great if application will promt about self-signed certs, and make user accept it, like browser do.

**Comment by skid9000** [comment]
> I'll have a look into it then, or I'll at least add an option to disable certificate checking.

Hello, i'm bumping this issue cause i use a local CA to sign my certificates but it seems that finamp dosen't use Android local user CA base to check if it's trusted or not.

I do want to still use my own CA so without disabling certificate checking, could that be possible ?

Thx !

**Comment by UnicornsOnLSD** [comment]
I believe this is an issue with Dart (or maybe BoringSSL?). I'll have a look into it because it's an annoying issue but no promises since poking around a programming language's source code isn't the easiest thing in the world.

**Comment by zuoez02** [comment]
Why not just add a checkbox on login screen? Just let user choose trust the server or not?

**Comment by ryneeverett** [comment]
Along the lines of the previous suggestion, I'd propose as an alternative to using the system's CAs, apps like this should allow permanent exceptions the way a browser does when you click through all the warnings. It's basically a TOFU model because it only adds an exception for that specific certificate. If you change the certificate you're back to the warning page, or in the case of this app you'd be back to the login page.

**Comment by avanier** [comment]
From what I can see even system certificates are not trusted. I'm trying to connect to a Jellyfin instance that has a known good self-signed certificate for which I have the root CA for the signing authority in my system bundle and I'm getting the handshake error. Is this the expected behaviour?

From the Flutter documentation it seems like CAs that are part of the system bundle should be respected. :confused:

**Comment by avanier** [comment]
Alright, uh, so mostly bad news. I've been digging down this :rabbit: :hole: and Flutter simply does not support using the system bundles at all since Chopper [does not respect `network-security-config`](https://github.com/flutter/flutter/issues/65841#issuecomment-692810524). It's a known bug. :disappointed: 

I tried building the app while injecting my root CA as an additional anchor ... and even that's not being respected. You can find sample code _(that cannot work)_ under this commit https://github.com/avanier/finamp/commit/229f038629d84b7aeef1cdef67d522b689325a1d . You'll see on the same branch as a drive-by I also tried updating Chopper to `5.0.0` and that didn't help.

I think the only realistic solution may be to add a checkbox to ignore the error like @zuoez02 suggests. Otherwise, to properly allow the use of system bundles would require getting rid of Chopper for something that makes native calls with the system API. Not a small task, and I'm not even sure it's possible at all under Flutter. _(Lowly sysadmin here, I know absolutely nothing of Android development.)_

**Comment by jmshrv** [comment]
Yeah...

It's a really weird omission from Dart, such a shame that we can't support it

**Comment by jmshrv** [comment]
I'll have to have a look to see how disabling certs worked, iirc you got a callback when something went wrong, so you might be able to override it for a specific cert.

Also, this isn't Chopper's fault, it's an issue in Dart. Otherwise, this would be a simple "move to a library that does support custom certificates"

**Comment by jmshrv** [comment]
Some relevant issues:

https://github.com/dart-lang/sdk/issues/48056
https://github.com/dart-lang/sdk/issues/50435

This function also looks interesting, we could have a manual option for now: https://api.dart.dev/stable/2.19.4/dart-io/SecurityContext/setTrustedCertificates.html

**Comment by jmshrv** [comment]
I'll get it sorted for the next release, which should *hopefully* be out this month. I've been busy with uni work since the last release, so I haven't been able to contribute much (a similar thing happened last year)

**Comment by jmshrv** [comment]
If `setTrustedCertificates` does what it sounds like, we could probably do it "properly" and have the certificate still be checked. It looks like the Flutter/Dart team are talking this seriously based on the activity in the issues linked [here](https://github.com/jmshrv/finamp/issues/106#issuecomment-1493452895), so it'll hopefully be even more proper soon :)

**Comment by jmshrv** [comment]
I'll have a look at both solutions, just don't want to accidentally make an insecure system. No point having a self-signed cert if someone can just MITM you by changing it

**Comment by jmshrv** [comment]
That's really weird. Maybe Xiaomi doesn't have the correct root certificate in the system store? Some browsers like to use their own root certificates.

When accessing from the local network, are you just connecting directly over HTTP?

**Comment by svalavuo** [comment]
Yeah. In local network I tested with http (without ssl) and local ip. I tried with three different browsers (Brave, Chrome and Mi Browser) and they all worked with https and my hostname. Finamp doesn't. :-(

**Comment by PlexSheep** [comment]
This issue is 2 years old by now. Can we please just get a checkbox that trusts this certificate?

**Comment by PlexSheep** [comment]
Yes. I understand it's a restriction with Google play. I'm personally using FDroid and I just can't use finamp when my servers certificate is not accepted.

I don't know how much of a maintenance effort it would be to do two versions, but it would be very handy. Especially regarding the fact that the used library/environment does not seem to address the issue.

**Comment by Chaphasilor** [comment]
Just to recap:

- Trusting custom certificates seems to be not possible. There are issues with the Flutter framework itself, and also with the HTTP client we're using
- Trusting *all* certificates might be possible, but I wouldn't bet on it. The native code we're using comes from other libraries, not us, and if they don't support it we're probably out of luck. I'm not sure how far @jmshrv got with his investigations...

I agree that adding a checkbox that explicitly disables all checks. This may or may not be supported by Google Play, but a custom build for download from GitHub should be entirely possible, once he issues above are solved.

**Comment by jmshrv** [comment]
The native code *may* use the system stores as intended, I haven't got around to this as I'm selfishly not affected by it myself

**Comment by PlexSheep** [comment]
> The native code *may* use the system stores as intended, I haven't got around to this as I'm selfishly not affected by it myself

Can you clarify what you mean by *may*?

**Comment by jmshrv** [comment]
I haven't actually checked but I'd assume ExoPlayer/AVPlayer would respect system certs as other native apps do

**Comment by jmshrv** [comment]
I can't find anything from Google saying that ignoring certificate errors isn't allowed. Browsers let you do it, after all

**Comment by Chaphasilor** [comment]
I'd argue a checkbox that's not checked by default and maybe hidden in an "advanced settings" menu is not a "vulnerability that exposes users to risk", it's a compatibility option.  
In any case, if it's not too much work to implement that handler, we should just try and see what Google says...

**Comment by Chaphasilor** [comment]
@OdinVex feel free to write the code, no need to use Android Studio. VS Code is all you need. But I don't think the allowlist and thumbprint detection is as easy as it seems. This is a cross-platform framework, almost nothing is consistent or easy to access when it comes to low-level functionality...

**Comment by Chaphasilor** [comment]
@OdinVex if you already have mostly working code, why not open a draft PR so we can take a look and get it ready to merge? :)

**Comment by Chaphasilor** [comment]
I'd be happy to help out with UI issues. Looking forward to the PR!

**Comment by nrdxp** [comment]
I use step-ca to provide an internal cert authority for services on my private (nebula based) mesh network. I install the root certificate to all the devices in the network, including my phone, to allow for apps like jellyfin to juse TLS connections without issue. This works on all apps except finamp. It works on the official jellyfin app, for example.

If I understand this thread correctly, this is because the Dart language does not have an interface to check against user installed certificates on android? Is there an open issue for this upstream? I'd be happy to ping that thread as well.

**Comment by spacekitteh** [comment]
Is any progress being made on this, especially for the FDroid version?

**Comment by Chaphasilor** [comment]
The Flutter app for Immich has recently added support for this, so we should be able to follow their approach. I've left a comment about it in #546. I don't use self-signed certs myself, so I'd appreciate someone with more experience in that regard to give it a go.

@OdinVex Finamp also runs on Linux now and I've heared of people using it on PostmarketOS/ARM. Just thought I'd let you know, in case you want to keep using it ^^

**Comment by Chaphasilor** [comment]
I guess we'll just wait a bit and see what happens to Immich then ^^  
If they don't get removed, as a much more popular app than Finamp and a direct competitor to Google Photos, than it might be worth the risk :)

**Comment by Chaphasilor** [comment]
If anyone want to try out the self-signed certificate support in Finamp, please use this APK test it:
https://gofile.io/d/jJ5P0w  
This test version will install right alongside your existing install of Finamp (if you have one), and won't mess with it. That means you'll have to sign into this version separately.  
Your installed self-signed certificates should automatically be picked up, so theoretically everything should just work™️

You can also download Windows and Linux (debug) builds for this [from here](https://github.com/jmshrv/finamp/actions/runs/16189369969), but those *will* interfere with the existing installs, meaning that you can't open both versions at the same time and you might have to delete your settings if we make any drastic changes before the next release.

Please let me know how it goes!

**Comment by jfly** [comment]
@Chaphasilor, should this be closed now that [my PR](https://github.com/jmshrv/finamp/pull/1276) has landed? Or do all these sorts of things (fixed in `redesign`, not yet in `main`) sit around open until the redesign is stable?

**Comment by Chaphasilor** [comment]
I'll close it after the changes are released to beta :)
